Composer/packages/lib/indexers/src/dialogUtils/dialogChecker.ts

```diff
@@ -11,30 +11,94 @@ import { CheckerFunc } from './types';
 
 const ExpressionParser = new ExpressionEngine();
 
-export const IsExpression: CheckerFunc = (
-  path,
-  value,
-  optional: { properties: string[]; requiredTypes: { [key: string]: boolean } }
-) => {
-  const { properties, requiredTypes } = optional;
-  return properties.reduce((result: Diagnostic[], property) => {
-    let message = '';
-    const exp = get(value, property);
-    if (!exp && requiredTypes[property]) {
-      message = formatMessage(`is missing or empty`);
-    } else {
-      try {
-        ExpressionParser.parse(exp);
-      } catch (error) {
-        message = `${formatMessage('must be an expression:')} ${error})`;
-      }
+const createPath = (path: string, type: string): string => {
+  const steps = ['triggers', 'actions', 'elseActions'];
+  let list = path.split('.');
+  const matchs = list.filter(x => !steps.every(step => !x.startsWith(step)));
+
+  const focused = matchs.join('.');
+  list = path.split(`${focused}.`);
+  if (list.length !== 2) return path;
+
+  return `${list[0]}${focused}#${type}#${list[1]}`;
+};
+
+export const checkExpression = (exp: string, required: boolean, path: string, type: string): Diagnostic | null => {
+  let message = '';
+  if (!exp && required) {
+    message = formatMessage(`is missing or empty`);
+  } else {
+    try {
+      ExpressionParser.parse(exp);
+    } catch (error) {
+      message = `${formatMessage('must be an expression:')} ${error})`;
     }
-    if (message) {
-      const diagnostic = new Diagnostic(message, '');
-      diagnostic.path = `${path}#${value.$type}#${property}`;
-      result.push(diagnostic);
+  }
+  if (message) {
+    const diagnostic = new Diagnostic(message, '');
+    diagnostic.path = createPath(path, type);
+    return diagnostic;
+  }
+
+  return null;
+};
+
+function findAllRequiredType(schema: any): { [key: string]: boolean } {
+  if (!schema) return {};
+  const types = schema.anyOf?.filter(x => x.title === 'Type');
+  const required = {};
+  if (types && types.length) {
+    types[0].required.forEach((element: string) => {
+      required[element] = true;
+    });
+  }
+
+  if (schema.required) {
+    schema.required.forEach((element: string) => {
+      required[element] = true;
+    });
+  }
+  return required;
+}
+
+export const IsExpression: CheckerFunc = (path, value, type, schema) => {
+  if (!schema) return [];
+  const diagnostics: Diagnostic[] = [];
+  const requiredTypes = findAllRequiredType(schema);
+  Object.keys(value).forEach(key => {
+    const property = value[key];
+    if (Array.isArray(property)) {
+      const itemsSchema = schema.properties[key].items;
+      if (itemsSchema?.$role === 'expression') {
```

Comment:
If we want to use the TS optional chaining operator we'll need to add a babel transform to support the browsers that Composer is to run in, right?

Reply:
Hi, @a-b-r-o-w-n has already updated the typescript's version to 3.7.2. This project use tsc to build. I think we don't need to use babel to transform.